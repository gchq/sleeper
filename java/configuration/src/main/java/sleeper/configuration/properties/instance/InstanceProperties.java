/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.configuration.properties.instance;

import com.amazonaws.services.s3.AmazonS3;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import sleeper.configuration.properties.SleeperProperties;
import sleeper.configuration.properties.SleeperPropertyIndex;
import sleeper.configuration.properties.format.SleeperPropertiesPrettyPrinter;

import java.io.IOException;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import java.util.Properties;

import static sleeper.configuration.properties.PropertiesUtils.loadProperties;
import static sleeper.configuration.properties.instance.CdkDefinedInstanceProperty.CONFIG_BUCKET;
import static sleeper.configuration.properties.instance.CommonProperty.TAGS;

/**
 * Contains all the properties needed to deploy an instance of Sleeper.
 */
public class InstanceProperties extends SleeperProperties<InstanceProperty> {
    private static final Logger LOGGER = LoggerFactory.getLogger(InstanceProperties.class);

    public static final String S3_INSTANCE_PROPERTIES_FILE = "instance.properties";

    protected Map<String, String> tags = new HashMap<>();

    public InstanceProperties() {
        super();
    }

    public InstanceProperties(Properties properties) {
        super(properties);
        tags = csvTagsToMap(get(TAGS));
    }

    public static InstanceProperties copyOf(InstanceProperties instanceProperties) {
        return new InstanceProperties(loadProperties(instanceProperties.saveAsString()));
    }

    @Override
    protected void init() {
        tags = csvTagsToMap(get(TAGS));
        super.init();
    }

    public Map<String, String> getTags() {
        return tags;
    }

    public void setTags(Map<String, String> tagsMap) {
        tags.clear();
        tags.putAll(tagsMap);
        set(TAGS, tagsToString(tags));
    }

    public void loadTags(Properties tagsProperties) {
        tags.clear();
        tagsProperties.stringPropertyNames().forEach(tagName -> tags.put(tagName, tagsProperties.getProperty(tagName)));
        if (!tags.isEmpty()) {
            set(TAGS, tagsToString(tags));
        }
    }

    @Override
    public String get(InstanceProperty property) {
        return compute(property, value -> property.computeValue(value, this));
    }

    public Properties getTagsProperties() {
        Properties tagsProperties = new Properties();
        tags.forEach(tagsProperties::setProperty);
        return tagsProperties;
    }

    public String getTagsPropertiesAsString() throws IOException {
        StringWriter stringWriter = new StringWriter();
        Properties tagsProperties = getTagsProperties();
        tagsProperties.store(stringWriter, "");
        return stringWriter.toString();
    }

    public static String getConfigBucketFromInstanceId(String instanceId) {
        return String.join("-", "sleeper", instanceId, "config").toLowerCase(Locale.ROOT);
    }

    public void loadFromS3GivenInstanceId(AmazonS3 s3Client, String instanceId) {
        String configBucket = getConfigBucketFromInstanceId(instanceId);
        loadFromS3(s3Client, configBucket);
    }

    public void loadFromS3(AmazonS3 s3Client, String bucket) {
        super.loadFromS3(s3Client, bucket, S3_INSTANCE_PROPERTIES_FILE);
    }

    public static Properties loadPropertiesFromS3GivenInstanceId(AmazonS3 s3Client, String instanceId) {
        return loadProperties(loadStringFromS3GivenInstanceId(s3Client, instanceId));
    }

    private static String loadStringFromS3GivenInstanceId(AmazonS3 s3Client, String instanceId) {
        String configBucket = getConfigBucketFromInstanceId(instanceId);
        return s3Client.getObjectAsString(configBucket, S3_INSTANCE_PROPERTIES_FILE);
    }

    public void saveToS3(AmazonS3 s3Client) {
        super.saveToS3(s3Client, get(CONFIG_BUCKET), S3_INSTANCE_PROPERTIES_FILE);
        LOGGER.info("Saved instance properties to bucket {}, key {}", get(CONFIG_BUCKET), S3_INSTANCE_PROPERTIES_FILE);
    }

    @Override
    public SleeperPropertyIndex<InstanceProperty> getPropertiesIndex() {
        return InstanceProperty.Index.INSTANCE;
    }

    @Override
    protected SleeperPropertiesPrettyPrinter<InstanceProperty> getPrettyPrinter(PrintWriter writer) {
        return SleeperPropertiesPrettyPrinter.forInstanceProperties(writer);
    }

    public static Map<String, String> csvTagsToMap(String csvTags) {
        Map<String, String> tags = new HashMap<>();
        if (null != csvTags && !csvTags.isEmpty()) {
            String[] split = csvTags.split(",");
            for (int i = 0; i < split.length; i += 2) {
                tags.put(split[i], split[i + 1]);
            }
        }
        return tags;
    }

    public static String tagsToString(Map<String, String> tags) {
        StringBuilder builder = new StringBuilder();
        int count = 0;
        for (Map.Entry<String, String> entry : tags.entrySet()) {
            if (count > 0) {
                builder.append(",");
            }
            builder.append(entry.getKey());
            builder.append(",");
            builder.append(entry.getValue());
            count++;
        }
        return builder.toString();
    }
}
