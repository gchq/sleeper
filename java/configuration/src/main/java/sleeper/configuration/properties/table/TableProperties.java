/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.configuration.properties.table;

import org.apache.commons.lang3.builder.EqualsBuilder;
import org.apache.commons.lang3.builder.HashCodeBuilder;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import sleeper.configuration.properties.SleeperProperties;
import sleeper.configuration.properties.SleeperPropertiesValidationReporter;
import sleeper.configuration.properties.SleeperPropertyIndex;
import sleeper.configuration.properties.format.SleeperPropertiesPrettyPrinter;
import sleeper.configuration.properties.instance.InstanceProperties;
import sleeper.core.schema.Schema;
import sleeper.core.schema.SchemaSerDe;
import sleeper.core.table.TableStatus;

import java.io.PrintWriter;
import java.util.Optional;
import java.util.Properties;

import static sleeper.configuration.properties.PropertiesUtils.loadProperties;
import static sleeper.configuration.properties.table.TableProperty.COMPACTION_FILES_BATCH_SIZE;
import static sleeper.configuration.properties.table.TableProperty.SCHEMA;
import static sleeper.configuration.properties.table.TableProperty.STATESTORE_CLASSNAME;
import static sleeper.configuration.properties.table.TableProperty.TABLE_ID;
import static sleeper.configuration.properties.table.TableProperty.TABLE_NAME;
import static sleeper.configuration.properties.table.TableProperty.TABLE_ONLINE;

public class TableProperties extends SleeperProperties<TableProperty> {
    private static final Logger LOGGER = LoggerFactory.getLogger(TableProperties.class);

    // Schema is cached for faster access
    private Schema schema;
    private final InstanceProperties instanceProperties; // Used for default properties

    public TableProperties(InstanceProperties instanceProperties) {
        this.instanceProperties = instanceProperties;
    }

    public TableProperties(InstanceProperties instanceProperties, Properties properties) {
        super(properties);
        this.instanceProperties = instanceProperties;
        schema = loadSchema(properties);
    }

    public static TableProperties copyOf(TableProperties tableProperties) {
        InstanceProperties instanceProperties = InstanceProperties.copyOf(tableProperties.instanceProperties);
        return new TableProperties(instanceProperties, loadProperties(tableProperties.saveAsString()));
    }

    public static TableProperties loadAndValidate(InstanceProperties instanceProperties, Properties properties) {
        TableProperties tableProperties = new TableProperties(instanceProperties, properties);
        tableProperties.validate();
        return tableProperties;
    }

    public static TableProperties reinitialise(TableProperties tableProperties, Properties newProperties) {
        return new TableProperties(tableProperties.instanceProperties, newProperties);
    }

    private static Schema loadSchema(Properties properties) {
        return Optional.ofNullable(properties.getProperty(SCHEMA.getPropertyName()))
                .map(Schema::loadFromString)
                .orElse(null);
    }

    @Override
    protected void init() {
        String schemaProperty = get(TableProperty.SCHEMA);
        if (schemaProperty != null) {
            schema = Schema.loadFromString(schemaProperty);
        }
        super.init();
    }

    @Override
    public void validate(SleeperPropertiesValidationReporter reporter) {
        super.validate(reporter);

        // This limit is based on calls to WriteTransactItems in DynamoDBFileReferenceStore.atomicallyUpdateX.
        // Also see the DynamoDB documentation:
        // https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/transaction-apis.html
        if ("sleeper.statestore.dynamodb.DynamoDBStateStore".equals(get(STATESTORE_CLASSNAME))
                && getInt(COMPACTION_FILES_BATCH_SIZE) > 49) {
            LOGGER.warn("Detected a compaction batch size for this table which would be incompatible with the " +
                    "chosen statestore. Maximum value is 49.");
            reporter.invalidProperty(COMPACTION_FILES_BATCH_SIZE, get(COMPACTION_FILES_BATCH_SIZE));
        }
    }

    @Override
    public String get(TableProperty property) {
        return compute(property, value -> property.computeValue(value, instanceProperties, this));
    }

    public Schema getSchema() {
        return schema;
    }

    public void setSchema(Schema schema) {
        this.schema = schema;
        set(TableProperty.SCHEMA, new SchemaSerDe().toJson(schema));
    }

    @Override
    public SleeperPropertyIndex<TableProperty> getPropertiesIndex() {
        return TableProperty.Index.INSTANCE;
    }

    @Override
    protected SleeperPropertiesPrettyPrinter<TableProperty> getPrettyPrinter(PrintWriter writer) {
        return SleeperPropertiesPrettyPrinter.forTableProperties(writer);
    }

    public TableStatus getStatus() {
        return TableStatus.uniqueIdAndName(get(TABLE_ID), get(TABLE_NAME), getBoolean(TABLE_ONLINE));
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        TableProperties that = (TableProperties) o;

        return new EqualsBuilder()
                .appendSuper(super.equals(o))
                .append(instanceProperties, that.instanceProperties)
                .isEquals();
    }

    @Override
    public int hashCode() {
        return new HashCodeBuilder(17, 37)
                .appendSuper(super.hashCode())
                .append(instanceProperties)
                .toHashCode();
    }
}
