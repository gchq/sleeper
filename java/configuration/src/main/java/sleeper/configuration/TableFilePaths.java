/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package sleeper.configuration;

import sleeper.core.partition.Partition;
import sleeper.core.properties.instance.InstanceProperties;
import sleeper.core.properties.table.TableProperties;

import static sleeper.core.properties.instance.CdkDefinedInstanceProperty.DATA_BUCKET;
import static sleeper.core.properties.instance.CommonProperty.FILE_SYSTEM;
import static sleeper.core.properties.table.TableProperty.TABLE_ID;

/**
 * Generator for paths to Sleeper table data files in the data bucket.
 */
public class TableFilePaths {

    private final String filePathPrefix;

    private TableFilePaths(String filePathPrefix) {
        this.filePathPrefix = filePathPrefix;
    }

    /**
     * Creates an instance of this class with the given prefix for paths. Paths will be generated assuming that this
     * prefix puts you at the base of the data bucket.
     *
     * @param  prefix the path prefix
     * @return        an instance of this class
     */
    public static TableFilePaths fromPrefix(String prefix) {
        return new TableFilePaths(prefix);
    }

    /**
     * Creates an instance of this class to generate full paths as they will be held in references to files in the state
     * store.
     *
     * @param  instanceProperties the instance properties
     * @param  tableProperties    the table properties
     * @return                    an instance of this class
     */
    public static TableFilePaths buildDataFilePathPrefix(InstanceProperties instanceProperties, TableProperties tableProperties) {
        return new TableFilePaths(instanceProperties.get(FILE_SYSTEM) + instanceProperties.get(DATA_BUCKET) + "/" + tableProperties.get(TABLE_ID));
    }

    /**
     * Generates a path to a data file in the given partition. Will have `.parquet` appended at the end. The filename
     * must match the filename used for the corresponding sketches file.
     *
     * @param  partition the partition
     * @param  fileName  the filename without file type
     * @return           the file path
     */
    public String constructPartitionParquetFilePath(Partition partition, String fileName) {
        return constructPartitionParquetFilePath(partition.getId(), fileName);
    }

    /**
     * Generates a path to a data file in the given partition. Will have `.parquet` appended at the end. The filename
     * must match the filename used for the corresponding sketches file.
     *
     * @param  partitionId the partition ID
     * @param  fileName    the filename without file type
     * @return             the file path
     */
    public String constructPartitionParquetFilePath(String partitionId, String fileName) {
        return String.format("%s/data/partition_%s/%s.parquet", filePathPrefix, partitionId, fileName);
    }

    /**
     * Generates a path to a sketches file in the given partition. Will have `.sketches` appended at the end. The
     * filename must match the filename used for the corresponding data file.
     *
     * @param  partition the partition
     * @param  fileName  the filename without file type
     * @return           the file path
     */
    public String constructQuantileSketchesFilePath(Partition partition, String fileName) {
        return String.format("%s/data/partition_%s/%s.sketches", filePathPrefix, partition.getId(), fileName);
    }

    public String getFilePathPrefix() {
        return filePathPrefix;
    }
}
