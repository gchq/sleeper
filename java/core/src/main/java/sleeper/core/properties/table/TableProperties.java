/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.properties.table;

import org.apache.commons.lang3.builder.EqualsBuilder;
import org.apache.commons.lang3.builder.HashCodeBuilder;

import sleeper.core.properties.PropertyGroup;
import sleeper.core.properties.SleeperProperties;
import sleeper.core.properties.SleeperPropertiesPrettyPrinter;
import sleeper.core.properties.SleeperPropertiesValidationReporter;
import sleeper.core.properties.SleeperPropertyIndex;
import sleeper.core.properties.instance.InstanceProperties;
import sleeper.core.schema.Schema;
import sleeper.core.schema.SchemaSerDe;
import sleeper.core.table.TableStatus;

import java.io.PrintWriter;
import java.util.Optional;
import java.util.Properties;
import java.util.stream.Collectors;

import static sleeper.core.properties.PropertiesUtils.loadProperties;
import static sleeper.core.properties.table.TableProperty.SCHEMA;
import static sleeper.core.properties.table.TableProperty.TABLE_ID;
import static sleeper.core.properties.table.TableProperty.TABLE_NAME;
import static sleeper.core.properties.table.TableProperty.TABLE_ONLINE;

/**
 * Contains values of the properties to configure a Sleeper table.
 */
public class TableProperties extends SleeperProperties<TableProperty> {

    // Schema is cached for faster access
    private Schema schema;
    private final InstanceProperties instanceProperties; // Used for default properties

    public TableProperties(InstanceProperties instanceProperties) {
        this.instanceProperties = instanceProperties;
    }

    public TableProperties(InstanceProperties instanceProperties, Properties properties) {
        super(properties);
        this.instanceProperties = instanceProperties;
        schema = loadSchema(properties);
    }

    /**
     * Creates a copy of the given table properties.
     *
     * @param  tableProperties the table properties
     * @return                 the copy
     */
    public static TableProperties copyOf(TableProperties tableProperties) {
        InstanceProperties instanceProperties = InstanceProperties.copyOf(tableProperties.instanceProperties);
        return new TableProperties(instanceProperties, loadProperties(tableProperties.saveAsString()));
    }

    /**
     * Creates and validates an instance of this class with the given property values.
     *
     * @param  instanceProperties the instance properties
     * @param  properties         the property values
     * @return                    the table properties
     */
    public static TableProperties createAndValidate(InstanceProperties instanceProperties, Properties properties) {
        TableProperties tableProperties = new TableProperties(instanceProperties, properties);
        tableProperties.validate();
        return tableProperties;
    }

    /**
     * Creates an instance with different property values than the given instance. Performs no validation, and uses the
     * same instance properties as the given instance.
     *
     * @param  tableProperties the original table properties
     * @param  newProperties   the property values
     * @return                 the new table properties
     */
    public static TableProperties recreateWithoutValidation(TableProperties tableProperties, Properties newProperties) {
        return new TableProperties(tableProperties.instanceProperties, newProperties);
    }

    private static Schema loadSchema(Properties properties) {
        return Optional.ofNullable(properties.getProperty(SCHEMA.getPropertyName()))
                .map(Schema::loadFromString)
                .orElse(null);
    }

    @Override
    protected void init() {
        String schemaProperty = get(TableProperty.SCHEMA);
        if (schemaProperty != null) {
            schema = Schema.loadFromString(schemaProperty);
        }
        super.init();
    }

    @Override
    public void validate(SleeperPropertiesValidationReporter reporter) {
        super.validate(reporter);
    }

    @Override
    public String get(TableProperty property) {
        return compute(property, value -> property.computeValue(value, instanceProperties, this));
    }

    public Schema getSchema() {
        return schema;
    }

    /**
     * Sets the schema of the Sleeper table.
     *
     * @param schema the schema
     */
    public void setSchema(Schema schema) {
        this.schema = schema;
        set(TableProperty.SCHEMA, new SchemaSerDe().toJson(schema));
    }

    @Override
    public SleeperPropertyIndex<TableProperty> getPropertiesIndex() {
        return TableProperty.Index.INSTANCE;
    }

    @Override
    protected SleeperPropertiesPrettyPrinter<TableProperty> getPrettyPrinter(PrintWriter writer) {
        return createPrettyPrinter(writer);
    }

    /**
     * Creates a printer to be used to display all table properties.
     *
     * @param  writer the writer to write to
     * @return        the pretty printer
     */
    public static SleeperPropertiesPrettyPrinter<TableProperty> createPrettyPrinter(PrintWriter writer) {
        return SleeperPropertiesPrettyPrinter.builder()
                .properties(TableProperty.getAll(), TablePropertyGroup.getAll())
                .writer(writer).build();
    }

    /**
     * Creates a printer to be used to display table properties in a given group.
     *
     * @param  writer the writer to write to
     * @param  group  the group to display
     * @return        the pretty printer
     */
    public static SleeperPropertiesPrettyPrinter<TableProperty> createPrettyPrinterWithGroup(
            PrintWriter writer, PropertyGroup group) {
        return SleeperPropertiesPrettyPrinter.builder()
                .sortedProperties(TableProperty.getAll().stream()
                        .filter(property -> property.getPropertyGroup().equals(group))
                        .collect(Collectors.toList()))
                .writer(writer).hideUnknownProperties(true).build();
    }

    public TableStatus getStatus() {
        return TableStatus.uniqueIdAndName(get(TABLE_ID), get(TABLE_NAME), getBoolean(TABLE_ONLINE));
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        TableProperties that = (TableProperties) o;

        return new EqualsBuilder()
                .appendSuper(super.equals(o))
                .append(instanceProperties, that.instanceProperties)
                .isEquals();
    }

    @Override
    public int hashCode() {
        return new HashCodeBuilder(17, 37)
                .appendSuper(super.hashCode())
                .append(instanceProperties)
                .toHashCode();
    }
}
