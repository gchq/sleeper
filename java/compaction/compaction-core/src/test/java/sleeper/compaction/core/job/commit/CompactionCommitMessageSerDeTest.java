/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.compaction.core.job.commit;

import org.approvaltests.Approvals;
import org.approvaltests.core.Options;
import org.junit.jupiter.api.Test;

import sleeper.core.partition.PartitionTree;
import sleeper.core.partition.PartitionsBuilder;
import sleeper.core.properties.instance.InstanceProperties;
import sleeper.core.properties.table.TableProperties;
import sleeper.core.schema.Schema;
import sleeper.core.statestore.FileReference;
import sleeper.core.statestore.FileReferenceFactory;
import sleeper.core.statestore.ReplaceFileReferencesRequest;
import sleeper.core.table.TableFilePaths;
import sleeper.core.table.TableIdGenerator;

import java.util.List;
import java.util.stream.IntStream;

import static org.assertj.core.api.Assertions.assertThat;
import static sleeper.core.properties.instance.CommonProperty.ID_MAX_LENGTH;
import static sleeper.core.properties.table.TableProperty.TABLE_ID;
import static sleeper.core.properties.table.TableProperty.TABLE_NAME;
import static sleeper.core.properties.testutils.InstancePropertiesTestHelper.createTestInstancePropertiesWithId;
import static sleeper.core.schema.SchemaTestHelper.schemaWithKey;
import static sleeper.core.testutils.SupplierTestHelper.exampleUUID;
import static sleeper.core.testutils.SupplierTestHelper.numberedUUID;

public class CompactionCommitMessageSerDeTest {

    InstanceProperties instanceProperties = createTestInstancePropertiesWithId("I".repeat(ID_MAX_LENGTH));
    Schema schema = schemaWithKey("key");
    String tableId = TableIdGenerator.fromRandomSeed(0).generateString();
    TableProperties tableProperties = createTestTableProperties(instanceProperties, schema, tableId);
    TableFilePaths filePaths = TableFilePaths.buildDataFilePathPrefix(instanceProperties, tableProperties);
    CompactionCommitMessageSerDe serDe = new CompactionCommitMessageSerDe();

    @Test
    void shouldSerialiseCompactionCommitRequest() {
        // Given
        PartitionTree partitions = new PartitionsBuilder(schemaWithKey("key")).singlePartition("root").buildTree();
        ReplaceFileReferencesRequest filesRequest = ReplaceFileReferencesRequest.builder()
                .jobId("test-job")
                .taskId("test-task")
                .jobRunId("test-run")
                .inputFiles(List.of("test.parquet"))
                .newReference(FileReferenceFactory.from(partitions).rootFile("output.parquet", 200))
                .build();
        CompactionCommitMessage message = new CompactionCommitMessage("test-table", filesRequest);
        Runnable callbackOnFail = () -> {
        };

        // When
        String json = serDe.toJson(message);
        CompactionCommitMessageHandle found = serDe.fromJsonWithCallbackOnFail(json, callbackOnFail);

        // Then
        assertThat(found).isEqualTo(new CompactionCommitMessageHandle("test-table", filesRequest, callbackOnFail));
    }

    @Test
    void shouldSerialiseRepresentativeCompactionCommitRequest() {
        // Given
        ReplaceFileReferencesRequest filesRequest = createExampleFilesRequestOfRepresentativeSize();
        CompactionCommitMessage message = new CompactionCommitMessage(tableId, filesRequest);
        Runnable callbackOnFail = () -> {
        };

        // When
        String json = serDe.toJson(message);
        CompactionCommitMessageHandle found = serDe.fromJsonWithCallbackOnFail(json, callbackOnFail);

        // Then
        assertThat(found).isEqualTo(new CompactionCommitMessageHandle(tableId, filesRequest, callbackOnFail));
    }

    @Test
    void shouldGenerateExpectedJson() {
        // Given
        CompactionCommitMessage message = new CompactionCommitMessage(tableId, createExampleFilesRequestOfRepresentativeSize());

        // When / Then
        Approvals.verify(serDe.toJson(message), new Options().forFile()
                .withName("example-compaction-commit", ".json"));
    }

    @Test
    void shouldGenerateExpectedPrettyJson() {
        // Given
        CompactionCommitMessage message = new CompactionCommitMessage(tableId, createExampleFilesRequestOfRepresentativeSize());

        // When / Then
        Approvals.verify(serDe.toJsonPrettyPrint(message), new Options().forFile()
                .withName("example-compaction-commit-pretty", ".json"));
    }

    private ReplaceFileReferencesRequest createExampleFilesRequestOfRepresentativeSize() {
        PartitionTree partitions = new PartitionsBuilder(schema).singlePartition(exampleUUID("partn", 0)).buildTree();
        List<String> inputFiles = IntStream.rangeClosed(1, 11)
                .mapToObj(i -> filePaths.constructPartitionParquetFilePath(
                        partitions.getRootPartition(), numberedUUID("file", i)))
                .toList();
        FileReference outputFile = FileReferenceFactory.from(partitions).rootFile(
                filePaths.constructPartitionParquetFilePath(
                        partitions.getRootPartition(), exampleUUID("file", 'N')),
                100_000_000);
        return ReplaceFileReferencesRequest.builder()
                .jobId(exampleUUID("job", 1))
                .taskId(exampleUUID("task", 1))
                .jobRunId(exampleUUID("run", 1))
                .inputFiles(inputFiles)
                .newReference(outputFile)
                .build();
    }

    private static TableProperties createTestTableProperties(InstanceProperties instanceProperties, Schema schema, String tableId) {
        TableProperties tableProperties = new TableProperties(instanceProperties);
        tableProperties.set(TABLE_ID, tableId);
        tableProperties.set(TABLE_NAME, "test");
        return tableProperties;
    }

}
