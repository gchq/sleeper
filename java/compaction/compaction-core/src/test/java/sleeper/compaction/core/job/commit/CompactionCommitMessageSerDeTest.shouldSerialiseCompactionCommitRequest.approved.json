{
  "tableId": "60b420bb",
  "request": {
    "jobId": "job11111-1111-1111-1111-111111111111",
    "taskId": "task1111-1111-1111-1111-111111111111",
    "jobRunId": "run11111-1111-1111-1111-111111111111",
    "inputFiles": [
      "s3a://sleeper-IIIIIIIIIIIIIIIIIIII-table-data/60b420bb/data/partition_partn000-0000-0000-0000-000000000000/file0000-0000-0000-0000-000000000000.parquet",
      "s3a://sleeper-IIIIIIIIIIIIIIIIIIII-table-data/60b420bb/data/partition_partn000-0000-0000-0000-000000000000/file1111-1111-1111-1111-111111111111.parquet",
      "s3a://sleeper-IIIIIIIIIIIIIIIIIIII-table-data/60b420bb/data/partition_partn000-0000-0000-0000-000000000000/file2222-2222-2222-2222-222222222222.parquet",
      "s3a://sleeper-IIIIIIIIIIIIIIIIIIII-table-data/60b420bb/data/partition_partn000-0000-0000-0000-000000000000/file3333-3333-3333-3333-333333333333.parquet",
      "s3a://sleeper-IIIIIIIIIIIIIIIIIIII-table-data/60b420bb/data/partition_partn000-0000-0000-0000-000000000000/file4444-4444-4444-4444-444444444444.parquet",
      "s3a://sleeper-IIIIIIIIIIIIIIIIIIII-table-data/60b420bb/data/partition_partn000-0000-0000-0000-000000000000/file5555-5555-5555-5555-555555555555.parquet",
      "s3a://sleeper-IIIIIIIIIIIIIIIIIIII-table-data/60b420bb/data/partition_partn000-0000-0000-0000-000000000000/file6666-6666-6666-6666-666666666666.parquet",
      "s3a://sleeper-IIIIIIIIIIIIIIIIIIII-table-data/60b420bb/data/partition_partn000-0000-0000-0000-000000000000/file7777-7777-7777-7777-777777777777.parquet",
      "s3a://sleeper-IIIIIIIIIIIIIIIIIIII-table-data/60b420bb/data/partition_partn000-0000-0000-0000-000000000000/file8888-8888-8888-8888-888888888888.parquet"
    ],
    "newReference": {
      "filename": "file9999-9999-9999-9999-999999999999",
      "partitionId": "partn000-0000-0000-0000-000000000000",
      "numberOfRecords": 200,
      "countApproximate": false,
      "onlyContainsDataForThisPartition": true
    }
  }
}